stacks-network/sbtc
# feat: rework stacks client to use only stacks nodes

## Description

Closes https://github.com/stacks-network/sbtc/issues/227. That ticket was referenced as the main follow-up work from https://github.com/stacks-network/sbtc/pull/228.

The https://github.com/stacks-network/sbtc/pull/228 PR added a stacks client trait for interacting with the stacks network. The one thing that was missing from that PR was a Nakamoto test. This PR changes the Stacks client trait from https://github.com/stacks-network/sbtc/pull/228 after seeing how Nakamoto actually works and adds a test.

## Changes

This PR has the following changes:
1. Change the function in the Stacks client trait to fetch all Nakamoto blocks for the previous tenure. Although I don't think this is the final API for this trait, I feel that it's closer than the previous iteration[^1].
2. Remove the dependency of the Hiro API[^2]. Hopefully we can keep it this way.
3. Simplify some of the error variants. Some of the fields in the variants were redundant.
4. Update the stacks-core dependency. This new branch includes most of the recent Nakamoto changes.
5. Add a test. This is just a unit test using a test fixture. It'll take a lot of fanfare to get Nakamoto running in CI, so this will have to do for now.


## Testing information

There is a detailed comment on how the unit test was setup, but the spark-notes are:
1. Run the `feat/signer` branch of the [stacks-regtest-env](https://github.com/hirosystems/stacks-regtest-env/tree/feat/signer) repo using a stacks node built on commit https://github.com/stacks-network/stacks-core/commit/3d96d53b35409859ca2baa2f0b6ddaa1fbd80265.
2. Make the API calls and check that things work!


[^1]: The old version of the trait used the bitcoin block hash. When there is a new bitcoin block there should be a tenure change (except for flash blocks?), so calling this new function will be functionally equivalent to the old one. We still need to be careful about data races between tenure changes and knowledge of new bitcoin blocks, but that isn't anything new with the new version here.
[^2]: The stacks-node Nakamoto endpoints did not recognize any of the block IDs returned from the Hiro API. Why? Not sure, but I suspect the current version of the Hiro API returns stacks block IDs that are different from what they'll be during Nakamoto. So I needed to avoid the Hiro API anyway.
